beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/duties/AttesterDutiesGenerator.java

```diff
@@ -0,0 +1,114 @@
+ */
+
+package tech.pegasys.teku.validator.coordinator.duties;
+
+import it.unimi.dsi.fastutil.ints.IntCollection;
+import it.unimi.dsi.fastutil.ints.IntList;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.spec.Spec;
+import tech.pegasys.teku.spec.config.SpecConfig;
+import tech.pegasys.teku.spec.datastructures.state.CommitteeAssignment;
+import tech.pegasys.teku.spec.datastructures.state.beaconstate.BeaconState;
+import tech.pegasys.teku.spec.logic.common.helpers.BeaconStateAccessors;
+import tech.pegasys.teku.spec.logic.common.helpers.MiscHelpers;
+import tech.pegasys.teku.validator.api.AttesterDuties;
+import tech.pegasys.teku.validator.api.AttesterDuty;
+
+public class AttesterDutiesGenerator {
+  private final Spec spec;
+
+  public AttesterDutiesGenerator(Spec spec) {
+    this.spec = spec;
+  }
+
+  public AttesterDuties getAttesterDutiesFromIndicesAndState(
+      final BeaconState state,
+      final UInt64 epoch,
+      final IntCollection validatorIndices,
+      final boolean isChainHeadOptimistic) {
+    final Bytes32 dependentRoot =
+        epoch.isGreaterThan(spec.getCurrentEpoch(state))
+            ? spec.atEpoch(epoch).getBeaconStateUtil().getCurrentDutyDependentRoot(state)
+            : spec.atEpoch(epoch).getBeaconStateUtil().getPreviousDutyDependentRoot(state);
+    final List<AttesterDuty> duties = createAttesterDuties(state, epoch, validatorIndices);
+    return new AttesterDuties(isChainHeadOptimistic, dependentRoot, duties);
+  }
+
+  private List<AttesterDuty> createAttesterDuties(
+      final BeaconState state, final UInt64 epoch, final IntCollection validatorIndices) {
+    final List<Optional<AttesterDuty>> maybeAttesterDutyList = new ArrayList<>();
+    final BeaconStateAccessors beaconStateAccessors = spec.atEpoch(epoch).beaconStateAccessors();
+    final UInt64 committeeCountPerSlot =
+        beaconStateAccessors.getCommitteeCountPerSlot(state, epoch);
+    final Map<Integer, CommitteeAssignment> validatorIndexToCommitteeAssignmentMap =
+        getValidatorIndexToCommitteeAssignmentMap(
+            state, beaconStateAccessors, committeeCountPerSlot.intValue(), epoch);
+    for (final Integer validatorIndex : validatorIndices) {
+      final CommitteeAssignment committeeAssignment =
+          validatorIndexToCommitteeAssignmentMap.get(validatorIndex);
+      if (committeeAssignment != null) {
+        maybeAttesterDutyList.add(
+            attesterDutyFromCommitteeAssignment(
+                committeeAssignment, validatorIndex, committeeCountPerSlot, state));
+      }
+    }
+    return maybeAttesterDutyList.stream().filter(Optional::isPresent).map(Optional::get).toList();
+  }
+
+  private Map<Integer, CommitteeAssignment> getValidatorIndexToCommitteeAssignmentMap(
```

Comment:
maybe moving this to `ValidatorsUtil` and access the function via `spec` as usual? so we have this next to`ValidatorsUtil::getCommitteeAssignment`?